plutus-core/src/Language/PlutusCore/Parser.y

```diff
@@ -190,21 +170,51 @@ getStaticBuiltinName = \case
     "ifThenElse"               -> Just IfThenElse
     _                          -> Nothing
 
-mkBuiltin :: a -> T.Text -> Term TyName Name uni a
-mkBuiltin loc ident = 
-   case getStaticBuiltinName ident of 
-      Just b  -> Builtin loc $ StaticBuiltinName b
-      Nothing -> Builtin loc (DynBuiltinName (DynamicBuiltinName ident))
-
--- FIXME: at this point it would be good to have access to the current
--- dynamic builtin names to check if a builtin with that name exists
--- and issue an error if not.  Unfortunately that involves adding a
--- parameter of type DynamicBuiltinNameMeanings to almost every
--- function in PlutusPrelude.hs, and others elsewhere.  @effectfully
--- says that his future plans may help with this.  In the meantime,
--- you get a scope resolution error during typechecking/execution if
--- it encounters a nonexistent name.  [We could also have a hard-coded
--- list of known dynamic names here, but that might not be ideal.]
+
+--- Parsing built-in types and constants ---
+
+-- | Tags of types in the default universe.
+tagOfTyName :: T.Text -> Maybe Int
+tagOfTyName = \case
+    "bool"       -> Just $ tagOf DefaultUniBool
+    "bytestring" -> Just $ tagOf DefaultUniByteString
+    "char"       -> Just $ tagOf DefaultUniChar
+    "integer"    -> Just $ tagOf DefaultUniInteger
+    "string"     -> Just $ tagOf DefaultUniString
+    "unit"       -> Just $ tagOf DefaultUniUnit
```

Comment:
So how about that idea to enumerate all types in a universe and construct a `Map` from pretty-printed type tags to, well, type tags? We could perhaps use [`inverseMap`](https://kowainik.github.io/posts/haskell-mini-patterns#bidirectional-parsing)

```haskell
inverseMap
    :: forall e s . (Bounded e, Enum e, Ord s)
    => (e -> s)
    -> s
    -> Maybe e
```

by having `e ~ Some (TypeIn uni)`. Then we only need to provide

```haskell
instance Enum (Some (TypeIn DefaultUni))
instance Bounded (Some (TypeIn DefaultUni))
```

to use the general parsing machinery for the default universe.

Reply:
I like that, but I think we can defer it until the issues for getting rid of static types and functions.  We should somehow pass the parser information about builtins so that nothing's actually coded into it.  I think this is a reasonable halfway point: we can merge it and fix it in a later PR (quite soon, hopefully).

Reply:
Fine by me.